Fix issues displaying polygons with holes in Shapes

jni opened the pull request:
# References

~~Partial fix for~~ Closes #5673. (!!!)
Closes #1653. 🥳 
Closes #1058. 🥳 
Closes #7799.

Depends on #7622 

# Description

(Updated 29/2/2024.)
(Updated 16/4/2025.)

This PR adds three preprocessing steps to polygon data during meshing in Shapes layers:
- discard duplicate polygon vertices, and recast edges in terms of unique vertices.
- discard duplicate edges — these are found when going to interior polygons depicting holes and when coming back.
- remove triangles from the triangulation that are not in the polygon. (only for the `triangle` backend, which otherwise includes these triangles.)

For the VisPy step, we go directly into the Triangulation class, because the PolygonData class completely ignores input edges when building a triangulation.

## Current limitations:

- ~~only works when `triangle` is installed (`py-triangle` in conda-forge)~~ **Edit:** nope, works with plain VisPy too!
- ~~currently on macOS, I am getting random segfaults when calling triangle.triangulate. I don't know what triggers them, but I suspect we're just going to have to implement our own triangulation, maybe with Numba.~~ **Edit:** these were caused by repeated vertices, against which triangulation algorithms are generally not robust. Uniquifying the vertices fixed the problem!
- ~~I still need to filter out the edges of the polygon for drawing, and maybe also for highlighting. Currently, the edges connecting the interior and exterior polygons are visible, which looks crap.~~ **Edit:** this is now done!
- ~~the polygons are processed twice, once for the interiors and once for the edges.~~ **Edit:** fixed by @Czaki 🥳 
- ~~the processing is probably a fair bit of overhead when many datasets would get no benefit. It might be worthwhile adding a flag to turn these checks on/off.~~ **Edit:** fixed by @Czaki 🥳 

## Summary of the issue and strategies:

It turns out that what we actually need is a *constrained Delaunay triangulation*[1], that is, a Delaunay triangulation that preserves specific edges from the vertices. That’s what the 'p' option in `triangle.triangulate` does *but* it doesn’t work if the edges are actually not planar — that’s why the triangulation was failing for this polygon even if triangle was installed.

This PR fixes that by finding edges that appear twice in the polygon and removing them.

Additionally, certain triangles from the resulting triangulation need to be removed, and this is easy to do by running `skimage.measure.points_in_poly` on the triangle centers, which are fast to compute with NumPy indexing.

[1]: http://en.wikipedia.org/wiki/Constrained_Delaunay_triangulation

jni committed:
Discard duplicate edges and interior triangles

This is a partial fix for #5673. Current limitations:

- only works when `triangle` is installed (`py-triangle` in conda-forge)
- currently on macOS, this works fine running in IPython but it segfaults when called from plain Python. I don’t know whether it’s just an env thing.

Summary:

It turns out that what we actually need is a *constrained Delaunay triangulation*[1], that is, a Delaunay triangulation that preserves specific edges from the vertices. That’s what the 'p' option in `triangle.triangulate` does *but* it doesn’t work if the edges are actually not planar — that’s why the triangulation was failing for this polygon even if triangle was installed. This PR fixes that by finding edges that appear twice in the polygon and removing them.

Additionally, certain triangles from the resulting triangulation need to be removed, and this is easy to do by running `skimage.measure.points_in_poly` on the triangle centers, which are fast to compute with NumPy indexing.

[1]: http://en.wikipedia.org/wiki/Constrained_Delaunay_triangulation

jni committed:
Add example of a polygon with a hole in it

jni committed:
Fix triangulation segfault

The issue was that the triangulation input cannot have repeated vertices. I fixed it by removing repeated vertices and translating all index coordinates to the unique vertices. This involves a bit more bookkeeping but it’s fine!

jni committed:
Update demo script to demonstrate the triangulation in matplotlib

jni committed:
Improve docs and robustness of dev script

jni committed:
Merge branch 'main' into polyhole

jni committed:
Add comments to make _generate_triangle_constraints clearer

jni committed:
Fix triangulation even in the case of VisPy

jni committed:
Refactor edge uniqueness function to be more general

jni committed:
PEP257 on two docstrings

jni committed:
Merge branch 'main' into polyhole

jni committed:
Ensure we norm with close=True for closed polygons

jni committed:
Initial working code fixing edge triangulation

jni committed:
Remove holes from path in separate function

jni committed:
Add some edge width to dev example for clarity

jni committed:
Use screenshot test to check hole is correctly rendered

jni committed:
Fix pixel check location

jni committed:
Don't try to remove holes before you have enough edges

jni committed:
Add docstring to _remove_holes

jni committed:
Merge branch 'main' into polyhole

jni committed:
Merge branch 'main' into polyhole

jni committed:
Guard napari.run with __name__ == '__main__' clause

psobolewskiPhD commented on examples/dev/poly-hole.py:
```diff
@@ -0,0 +1,97 @@
+
+fig, ax = plt.subplots()
+pltri(res['vertices'], res['triangles'], mask=in_poly, ax=ax)
+plt.show()
```
```suggestion
fig.show()
```
This fixes the example for me locally, but I can't say I know why.

jni committed:
use fig.show instead of plt.show

Apparently this fixes things in some situations!

Co-authored-by: Peter Sobolewski <76622105+psobolewskiPhD@users.noreply.github.com>

melonora committed:
Merge branch 'main' into polyhole

Czaki committed:
restore convex speedup

Czaki committed:
Merge branch 'main' into polyhole

Czaki commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -598,22 +702,27 @@ def triangulate_face(data: npt.NDArray) -> tuple[npt.NDArray, npt.NDArray]:
+            vertices, raw_triangles, polygon_vertices
+        )
+    elif _is_convex(polygon_vertices):
+        vertices, triangles = _fan_triangulation(polygon_vertices)
```
I prefer to call `_is_convex` before `_normalize_vertices_and_edges` as we do not use its output in `_fan_triangulation`

jni replied:
Yes that makes sense!

jni committed:
First check whether polygon is convex for faster and more robust triangulation

jni committed:
Add function to combine multiple triangulations

jni committed:
Keep edges around holes in polygons with holes

jni committed:
Merge branch 'main' into polyhole

jni committed:
Fix typing

jni committed:
Bug fix: don't check for holes with only two vertices

jni committed:
Ensure edge width is 0 since inner edges are drawn now

Czaki committed:
Merge branch 'main' into polyhole

Czaki commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -603,22 +707,27 @@ def triangulate_face(data: npt.NDArray) -> tuple[npt.NDArray, npt.NDArray]:
-        vertices, triangles = _fan_triangulation(data)
+    raw_vertices, edges = _normalize_vertices_and_edges(
+        polygon_vertices, close=True
+    )
```
We do not need to perform this operation before checking if the polygon is convex.

jni replied:
Fixed

Czaki replied:
Did you check that uint32 cast is obsolete?

jni replied:
Yes. You specifically create a unit32 array for the fan triangulation.

Czaki commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -603,22 +707,27 @@ def triangulate_face(data: npt.NDArray) -> tuple[npt.NDArray, npt.NDArray]:
+        )
+        vertices = res['vertices']
+        raw_triangles = res['triangles']
+        triangles = _cull_triangles_not_in_poly(
+            vertices, raw_triangles, polygon_vertices
+        )
```
Why do we need to apply this only to triangulate?

jni replied:
Triangle and vispy have different algorithms, and the vispy one doesn't include these triangles from the get-go. (I haven't actually studied the algorithms, I've just noticed this.)

Czaki replied:
Could you leave proper comment in code

jni replied:
Done

Czaki commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -699,13 +809,138 @@ def _sign_cross(x, y):
+    triangles : np.ndarray of int
+        Px3 array of indices into centers to form the triangles.
+    """
+    centers_list, offsets_list, triangles_list = list(zip(*meshes_list))
```
If len on the list is 1, we should just return it without calling concat.

jni replied:
Done!

Czaki replied:
Where? Did you not pushed some changes?

Czaki reviewed (dismissed):
As I work on performance improvement of polygons, I have checked this PR in this context. 
And it does not look good. 

Could we introduce a new Shape class (maybe `PolygonWithHoles`) and activate, a new code path only when shape belong to this class?

Czaki commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -732,6 +971,16 @@ def generate_2D_edge_meshes(path, closed=False, limit=3, bevel=False):
+                )
+                for p in _remove_internal_edges(path)
+            ]
+        )
```
If there is no hole, then calling `_remove_internal_edges` is expensive. However, we should know if it is required after `triangulate_face` call

jni replied:
Do you have a suggestion for keeping track of this? I noted this already [here](https://github.com/napari/napari/pull/6654#issuecomment-2061416901) and [here](https://github.com/napari/napari/pull/6654#issuecomment-2187923199), but I don't see a clean way to implement the check only once, because the face and edge triangulations happen surprisingly far from each other in the code...

jni committed:
Don't sanitise poly nodes and edges if the poly is convex

jni committed:
add comment about triangle culling

jni committed:
Return fast when only one mesh

jni committed:
Merge branch 'main' into polyhole

jni committed:
Merge branch 'main' into polyhole

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni committed:
formatting after zip fix

jni committed:
Remove redundant duplicates check in face triangulation

jni committed:
Simplify set_meshes_py, support planar axes other than 0

jni committed:
Ignore typing call

Not sure how else to fix this; we could return dummy values for
axis and value, but that seems worse than ignoring the type.

jni committed:
Alternative to ignoring type (and actually correct 😅)

jni committed:
Merge branch 'refactor-face' into polyhole

When #7622 is merged to main and main is merged here, these commits
should disappear.

Czaki committed:
Merge remote-tracking branch 'upstream/main' into polyhole

Czaki committed:
Merge remote-tracking branch 'upstream/main' into polyhole

Czaki committed:
Merge branch 'main' into polyhole

jni committed:
Replace duck-typing check with isinstance check for QRangeSlider, v2 (#7719)

In #7712, I missed one more "hasattr" check for a private attribute.
This one replaces it with an isinstance check.

Czaki committed:
implement removing edges

Czaki committed:
fix tests

Czaki committed:
improve typing

Czaki committed:
fix bug

Czaki committed:
fix typing

Czaki committed:
fix case catched by example

psobolewskiPhD committed:
[CI] Update test_vispy_camera.py to skip orientation_2d test on windows (#7717)

# References and relevant issues
Despite the previous attempt https://github.com/napari/napari/pull/7709
still getting windows seg faults:
https://github.com/napari/napari/issues/7715

# Description
Skip the other orientation test on windows.
Hopefully this solves it?
🤞

jni committed:
Merge remote-tracking branch 'origin/polyhole' into polyhole

jni commented on napari/layers/shapes/_accelerated_triangulate_dispatch.py:
```diff
@@ -256,6 +265,90 @@ def create_box_from_bounding_py(bounding_box: npt.NDArray) -> npt.NDArray:
+) -> list[CoordinateArray3D]: ...
+
+
+def reconstruct_polygon_edges_py(
```
@Czaki I don't think this implementation is necessary, since i already implemented it using SciPy sparse graphs? (ie in the absence of numba, let's use SciPy+NumPy.)

Czaki replied:
When I try your code on example it looks like this:
![obraz](https://github.com/user-attachments/assets/d0897352-5738-4b2a-b4e6-c1f495dc49f1)

Based on description it try only to return external edges that is not correct.

jni replied:
It was an arbitrary choice to only return external ones — I can do all polys easily.

jni replied:
My main point is that I have edge removal code already, using C routines from scipy. I'll be surprised if they benchmark slower than pure Python. And currently numba doesn't work.

jni commented on napari/layers/shapes/shape_types.py:
```diff
@@ -0,0 +1,17 @@
+
+CoordinateArray2D = np.ndarray[tuple[int, Literal[2]], np.dtype[np.float32]]
+CoordinateArray3D = np.ndarray[tuple[int, Literal[3]], np.dtype[np.float32]]
+CoordinateArray = CoordinateArray2D | CoordinateArray3D
```
Even though it's correct for our narrow use case, I don't like this — I think coordinate arrays should be able to have any dimensionality, and any float type. Are we able to turn it into a generic type? `CoordinateArray[Literal[2], np.dtype[np.float32]]`, `D = TypeVar(); CoordinateArray[D, np.float32]`, etc...

Czaki replied:
It was big pain to type existing code. 
I think that we should have different typing for rendering (only 2d and 3d) and different for general use case. Maybe different name could be better?

Czaki committed:
skip problematic benchmarks

Czaki committed:
skip more slow tests

Czaki committed:
Merge branch 'main' into polyhole

Czaki committed:
improve typing and skip coverage of excption of bug in triangulation

Czaki committed:
local version

Czaki committed:
check pure python

Czaki committed:
skip problematic case

Czaki committed:
use numba

Czaki committed:
cleanup

Czaki committed:
fix mypy

Czaki committed:
fix import

Czaki committed:
fix example

Czaki committed:
use numba for convex checking

Czaki committed:
add missed tests

Czaki committed:
fix typing

Czaki committed:
improve import for easier benchmarking

Czaki committed:
fix example

Czaki committed:
add functions to warmup numba

Czaki committed:
add more tests

Czaki committed:
ad numba reconstruct

Czaki committed:
more tests

Czaki committed:
Merge remote-tracking branch 'upstream/main' into polyhole

Czaki committed:
skip slow cases

Czaki committed:
skip triangle backend wihout holes fix.

jni commented on napari/benchmarks/benchmark_shapes_layer.py:
```diff
@@ -385,27 +385,31 @@ class _ShapeTriangulationBaseShapeCount(_ShapeTriangulationBase):
+        == 128
         and shape_type == 'polygon'
-        and compiled_triangulation != BackendType.triangle,
+        and backend_type != BackendType.triangle,
```
I would rename this `triangulation_backend`

Czaki replied:
I do this in #7747

jni commented on napari/benchmarks/benchmark_shapes_layer.py:
```diff
@@ -385,27 +385,31 @@ class _ShapeTriangulationBaseShapeCount(_ShapeTriangulationBase):
-        and compiled_triangulation != BackendType.triangle,
+        and backend_type != BackendType.triangle,
         if_in_pr=skip_above_100,
+        # to slow (40 sec)
```
```suggestion
        # too slow (40 sec)
```

Czaki replied:
Let's wait for merge until CI (pypi) starts working again

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+    for i in range(len(vertices)):
+        vertices_sum[0] += vertices[i][0]
+        vertices_sum[1] += vertices[i][1]
+    return vertices_sum / len(vertices)
```
(1) what is the performance impact of using this instead of `np.mean(vertices, axis=0)`?
(2) This algorithm is not numerically stable. Should we use the [numerically-stable computation](https://en.wikipedia.org/wiki/Moving_average#Cumulative_average) instead? It is much more expensive, but at least, maybe we should leave a note in case we run into weird edge cases in the future... 😅

Czaki replied:
You mean this algorithm:
https://en.m.wikipedia.org/wiki/Kahan_summation_algorithm ?

Czaki replied:
For `np.mean(vertices, axis=0)`, numba does not support such a call:

```
>           raise e.with_traceback(None)
E           numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
E           No implementation of function Function(<function mean at 0x733dc11f9580>) found for signature:
E            
E            >>> mean(array(int64, 2d, C), axis=Literal[int](0))
E            
E           There are 2 candidate implementations:
E                 - Of which 2 did not match due to:
E                 Overload in function 'array_mean': File: numba/np/old_arraymath.py: Line 424.
E                   With argument(s): '(array(int64, 2d, C), axis=int64)':
E                  Rejected as the implementation raised a specific error:
E                    TypingError: got an unexpected keyword argument 'axis'
E             raised from /home/czaki/.pyenv/versions/napari_3.11/lib/python3.11/site-packages/numba/core/typing/templates.py:791
E           
E           During: resolving callee type: Function(<function mean at 0x733dc11f9580>)
E           During: typing of call at /home/czaki/Projekty/napari/napari/layers/shapes/_accelerated_triangulate.py (472)
```

Czaki replied:
Removed `_centroid` and used `np.mean` twice.

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+    vertices: CoordinateArray2D, centroid: np.ndarray
+) -> bool:
+    """Check if the polygon, that all angles have the same orientation
+    do not have self-intersection
```
I'm not sure I understand this function... Can you explain what condition is actually being checked? e.g. this is my attempt:

```suggestion
    """Check that the polygon defined by `vertices` does not self-intersect.
    
    This function assumes that all the angles in the polygon have the same
    orientation. It then checks that all angles between the centroid and
    start vertex and the centroid and consecutive vertices are
    monotonically increasing — looping back past 360º implies a
    self-intersecting shape.
    
    Both conditions together (all internal angles have the same orientation
    and all angles to the centroid are monotonically increasing) imply a
    convex shape, which is fast to draw.
```

Czaki replied:
Hm. May we move this comment to `is_convex`. And here just left 

```python 
"""Check if vertices are monotical in polar system.

The function checks if in polar coordinate system
with point 0 in centroid, the angle component
of vertices coordinate is increasing.

This is a second part of polygon convexity check.
"""
```

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+    bool
+        True if the polygon is convex, False otherwise
+    """
+    # A polygon is convex if all its internal angles are less than 180 degrees.
```
and its edges don't self-intersect!

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+            triangle_orientation != Orientation.collinear
+            and triangle_orientation != orientation_
+        ):
+            return False
```
```suggestion
    original_orientation = Orientation.collinear
    # first, find the first non-collinear orientation
    for idx0 in range(vertices.shape[0] - 2):
        p1 = vertices[idx0]
        p2 = vertices[idx0 + 1]
        p3 = vertices[idx0 + 2]
        current_orientation = _orientation(p1, p2, p3)
        if current_orientation != Orientation.collinear:
            original_orientation = current_orientation
            break
    if current_orientation == Orientation.collinear:
        # we went through all vertices and all orientations were collinear:
        # the "polygon" is a line, and not a convex polygon
        return False
    # next, check all vertex triplets until just before looping around
    for idx in range(idx0 + 1, vertices.shape[0] - 2):
        p1 = vertices[idx]
        p2 = vertices[idx + 1]
        p3 = vertices[idx + 2]
        current_orientation = _orientation(p1, p2, p3)
        if (
            current_orientation != Orientation.collinear
            and curent_orientation != original_orientation
        ):
            # change in orientation implies non-convex
            return False
```

jni replied:
I prefer a for-loop, and also added comments to explain why the repeated code.

Czaki replied:
I always have a problem with leaking of idx variable of for loop. 

It is not how it works with other languages and is less readable form me. But if you prefer this.

jni replied:
yeah I also don't love it but I do prefer it to while-loops. Alternatively, I think it could be refactored as a single loop — have a flag to keep track of whether we have seen a non-collinear angle first, and if the flag is set, then we do the early exit. Let me try that. I can push to this branch.

Czaki replied:
No. This means additional if executed each loop. It is slower.

jni replied:
true, but the code is *much* nicer, and I doubt it's a big difference (1 bool flag vs loop condition + 3 array deref + 3 subtractions + 2 multiplications) *and* if people *really* want speed they will use bermuda, right? ;)

```python
    orientation_set = False
    n = vertices.shape[0]

    # we check all non-looping triplets first to avoid expensive
    # modulus operation in the full loop
    for idx in range(n - 2):
        p1 = vertices[idx]
        p2 = vertices[idx + 1]
        p3 = vertices[idx + 2]
        current_orientation = _orientation(p1, p2, p3)
        if current_orientation != Orientation.collinear:
            if not orientation_set:
                original_orientation = current_orientation
                orientation_set = True
            elif current_orientation != original_orientation:
                return False
```

Czaki replied:
changed

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+        triangle_orientation != Orientation.collinear
+        and triangle_orientation != orientation_
+    ):
+        return False
```
I can't suggest here because GH stupidly believes this code is related to the deleted line. But this is my suggestion:

```python
    # finally, check orientation when looping around to start of polygon
    n = vertices.shape[0]
    for idx in range(n - 2, n)
        p1 = vertices[idx]
        p2 = vertices[idx + 1 % n]
        p3 = vertices[idx + 2 % n]
        current_orientation = _orientation(p1, p2, p3)
        if (
            current_orientation != Orientation.collinear
            and curent_orientation != original_orientation
        ):
            # change in orientation implies non-convex
            return False
```

Czaki replied:
What is the benefit of for loop of length 2? It's just add modulo operation?

jni replied:
yes, I presumed that the reason you didn't make the complete loop is that you didn't want to run modulo always. I prefer the 2-loop to the code being duplicated, it makes it easier to know that we are just making the same check two times again.

Czaki replied:
Maybe `for (i1, i2, i3) in [(n-2, n-1, 0), (n-1, 0, 1)]` ?

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,139 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+    if orientation_ == Orientation.anticlockwise:
+        return _are_polar_angles_monotonic(vertices, centroid)
+
+    return _are_polar_angles_monotonic(vertices[::-1], centroid)
```
```suggestion
    # all the angles had the same orientation. We finally check that
    # there are no self intersections by checking the angles the
    # vertices make with the centroid.
    centroid = _centroid(vertices)

    if orientation_ == Orientation.anticlockwise:
        return _are_polar_angles_monotonic(vertices, centroid)

    return _are_polar_angles_monotonic(vertices[::-1], centroid)
```

Though tbh I would consider moving the orientation check inside the monotonic function:

```python
    # all the angles had the same orientation. We finally check that
    # there are no self intersections by checking the angles the
    # vertices make with the centroid.
    centroid = _centroid(vertices)

    return _are_polar_angles_monotonic(vertices, centroid, current_orientation)
```

Czaki replied:
Adding `current_orientation` to `_are_polar_angles_monotonic` is reducing readability for me.

jni reviewed (commented):
@Czaki I've made some comments, not yet done with full review but it's late here, I'll continue in the morning. But I think this is very very close! 😍

Czaki commented on napari/benchmarks/benchmark_shapes_layer.py:
```diff
@@ -445,31 +448,40 @@ class ShapeTriangulationStarIntersectionSuite(
+
+
+def skip_triangle_backend(_n_shapes, _n_points, _shape_type, backend_type):
+    """THis is a helper function to skip triangle testing in polygon with holes case if
```
```suggestion
    """This is a helper function to skip triangle testing in polygon with holes case if
```

Czaki committed:
Apply suggestions from code review

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>

Czaki committed:
improve is_convex docstring

Czaki committed:
simplify code

Czaki committed:
rename backend_type to triangulation_backend

Czaki committed:
simplify `is_convex` code

Czaki committed:
calculate centroid using np.mean

TimMonko commented on examples/dev/poly-hole.py:
Is there a way to unblock (easily) opening napari and mpl, currently, I get a really slow startup with the example and it makes it appear to be not working.

Czaki replied:
What is hard about opening both napari and mpl?

jni replied:
I think @TimMonko is saying that this example seems slow to launch compared to others. Let's figure that out later though. 😅 

One thing though is we now have an extra API that I think we should use — `viewer.camera.orientation2d = ('up', 'right')` — because this polygon is actually [South Africa](https://en.wikipedia.org/wiki/South_Africa#/media/File:South_Africa_adm_location_map.svg) upside down. 😂 (and remove the - sign from the coordinates for mpl)

jni replied:
(I pushed these changes)

TimMonko commented on napari/_tests/test_with_screenshot.py:
```diff
@@ -556,6 +556,37 @@ def test_blending_modes_with_canvas(make_napari_viewer):
     np.testing.assert_array_equal(screenshot[:, :, 0], np.minimum(img1, img2))
 
 
+@skip_on_win_ci
```
why skipping on windows ci? works for me locally

jni replied:
ah yes, and everybody knows that it working locally means it'll *definitely* work in CI... 😜 

Seriously though, the answer is I don't remember why I added this, but there's a good chance the windows build failed before it. (There's a few cases of screenshot tests not working on Windows.)

Alternatively, tests around this one all had the decorator and I decided to do as the Romans do.

Either way, I don't think we should touch while CI is working 😅 as it can always be tweaked later.

Czaki replied:
We may open a separate PR to test it after merging this PR.

TimMonko commented on napari/benchmarks/benchmark_shapes_layer.py:
```diff
@@ -27,7 +27,7 @@
 
 
-class BackendType(Enum):
+class TriangulationBackend(Enum):
```
not needed for this PR, but this would be a good thing to move to a utils file. From there we can reuse the enum to populate preference settings for the backend (instead of the checkbox as is)

jni replied:
I think the plan is to do this in #7747

Czaki replied:
The general plan is to allow control this by settings (or module level variable in napari.layers.shapes._shape_models.shapes), but until we release 0.6.0 we need more complex machinery, as benchmarks needs to work with the latest released version.

TimMonko commented on napari/benchmarks/benchmark_shapes_layer.py:
```diff
@@ -79,24 +83,22 @@ def _disable_numba(self):
                 )
 
-    def select_backend(self, triangulation_backend: BackendType):
+    def select_backend(self, triangulation_backend: TriangulationBackend):
```
wait, I guess at this point I'm confused why the backend is being selected in the benchmarks file? wouldn't this be more appropriate for a different module to more clearly separate the functionality of selected backend from benchmarking? (I guess this goes hand in hand with above Enum comment

jni replied:
I'm not sure about the exact implementation design, but changing it in benchmarks is so that, regardless of settings, we can check each PR for regression in *any* backend. See also #7739.

jni replied:
(and again, the settings issue is being addressed in #7747)

Czaki replied:
To produce such a nice table with measures for each backend (parts of checks are skipped for performance)

```
[50.00%] ··· benchmark_shapes_layer.ShapeTriangulationNonConvexSuite.time_create_layer                                                                                                  ok
[50.00%] ··· ========== ========== ==================== ================= ============== ======================= ==================== =================
             --                                                             shape_type / triangulation_backend
             --------------------- --------------------------------------------------------------------------------------------------------------------
              n_shapes   n_points   path / partsegcore   path / triangle   path / numba   polygon / partsegcore   polygon / triangle   polygon / numba
             ========== ========== ==================== ================= ============== ======================= ==================== =================
                100         8            243±0ms             302±0ms         283±0ms             5.15±0s               358±0ms             591±0ms
                100         32           321±0ms             286±0ms         315±0ms             1.20±0s               301±0ms             1.08±0s
                100        128           375±0ms             338±0ms         404±0ms               n/a                 424±0ms               n/a
                5000        8              n/a                 n/a             n/a                 n/a                   n/a                 n/a
                5000        32             n/a                 n/a             n/a                 n/a                   n/a                 n/a
                5000       128             n/a                 n/a             n/a                 n/a                   n/a                 n/a
             ========== ========== ==================== ================= ============== ======================= ==================== =================
```

TimMonko commented on napari/layers/shapes/_tests/conftest.py:
seems comprehensive enough to me, lets hope so!

TimMonko commented on napari/layers/shapes/_tests/test_shapes_utils.py:
how would polylines and paths be expected to interact with `is_convex_py`? what _should_ happen. Would a test be helpful?

TimMonko replied:
oh, maybe thats what `line` from conftest is? a comment would be nice if so! I at first assumed it was a line segment with multiple points in side, not multiple connected line segments

Czaki replied:
I do not understand the requested change. Could you rephrase?

TimMonko commented on napari/layers/shapes/_tests/test_triangulation.py:
```diff
@@ -8,8 +8,8 @@
-@pytest.fixture
-def _disable_jit(monkeypatch):
+@pytest.fixture(params=[False, True])
+def _disable_jit(monkeypatch, request):
```
```suggestion
def _disable_jit(request):
```
monkeypatch not used, I think? at least I don't see it + tests still pass for me

TimMonko commented on napari/layers/shapes/_shapes_models/shape.py:
new typings here definitely improved readability👍

TimMonko commented on napari/layers/shapes/_shapes_models/shape.py:
```diff
@@ -327,14 +339,27 @@ def _set_meshes_py(
+            self._edge_vertices = np.empty(
+                (0, self.ndisplay), dtype=np.float32
+            )
+            self._edge_offsets = np.empty((0, self.ndisplay), dtype=np.float32)
```
comments on why initializing with the specific dtypes would be nice
specially since triangles are uint32 and the new ones are float32

Czaki replied:
Triangles are still uint32. 
here we only changed vertices and offsets from default float (linux/macOS float64, windows float32) to float32, as thes detype is sent to vispy.

TimMonko commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,120 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+
+    Then we check if there are no self-intersections by checking if the
+    polar angles of the vertices are monotonic.
+    The center of a polar coordinate system is the centroid of the polygon.
```
```suggestion
    """Check if a polygon is convex.

    A polygon is convex when all its internal angles
    are less than or equal to 180 degrees and its edges don't
    self-intersect.
    
    This function determines convexity by:
    1. Checking if all non-collinear angles have the same
        orientation (clockwise or counterclockwise)
    2. Verifying that the vertices' polar angles are monotonic,
        relative to the polygon's centroid.
        
    - If the vertices are ordered counterclockwise, the order is
    reversed before checking.
    - Polygons with fewer than 3 vertices are not considered convex.
    - A triangle is always convex.
```

TimMonko commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,120 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+    Parameters
+    ----------
+    vertices : np.ndarray
+        Array of vertex coordinates with shape (N, 2)
```
```suggestion
        Array of vertex coordinates with shape (N, 2),
        where N is the number of vertices
```

TimMonko commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,120 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+def _are_polar_angles_monotonic(
+    vertices: CoordinateArray2D, centroid: np.ndarray
+) -> bool:
+    """Check if vertices are monastical in a polar system.
+
+    The function checks if in a polar coordinate system
+    with point 0 in centroid, the angle component
+    of vertices coordinate is increasing.
+
+    This is a second part of the polygon convexity check.
+
+    Parameters
+    ----------
+    vertices: np.ndarray
+        Array of vertex coordinates with shape (N, 2)
+
+    Returns
+    -------
+    bool
+        True if the polygon is simple, False otherwise
+    """
```
```suggestion
    """Check if polar angles of vertices are monotonic.

    This function checks whether the polar angles of the
    vertices, relative to a given centroid, increase monotonically
    in a polar coordinate system. If the angles are not monotonic,
    the polygon may have self-intersections or non-convex regions.
    This is the second part of the polygon convexity check. 
    
    
    
    The function checks if in a polar coordinate system
    with point 0 in centroid, the angle component
    of vertices coordinate is increasing.

    This is a second part of the polygon convexity check.

    Parameters
    ----------
    vertices: np.ndarray
        Array of vertex coordinates with shape (N, 2),
        where N is the number of vertices
    centroid: np.ndarray
        The centroid of the polygon, used as the origin of the polar
        coordinate system.

    Returns
    -------
    bool
        True if the polygon is simple (polar angles are monotonically
        increasing), False otherwise.
    """
```

TimMonko commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -319,11 +342,120 @@ def _orientation(p1: np.ndarray, p2: np.ndarray, p3: np.ndarray) -> float:
+            - start_angle
+        )
+        if angle < 0:
+            angle += 2 * pi
```
why do angles need to be positive? or is it more just that for the next line we are really checking  the absolute value of current and last angle against each other? I'm doing my best to understand, you are doing great work!

jni replied:
This is how I understand it: the result of [atan2](https://docs.python.org/3/library/math.html#math.atan2) is in [-π, π]. We are trying to check if, when we trace out the vertices, the angle is always increasing until we come back full circle. To do that, we want to get angles in [0, 2π]. So as soon as we flip past 180º, we add 2π ([=τ](https://www.tauday.com)) so we can keep going around the circle.

Czaki replied:
As @jni wrote

TimMonko commented on napari/layers/shapes/_accelerated_triangulate_dispatch.py:
```diff
@@ -256,6 +265,283 @@ def create_box_from_bounding_py(bounding_box: npt.NDArray) -> npt.NDArray:
+
+    This function ensures that:
+
+    - no nodes are repeated, as this can cause problems with triangulation
+      algorithms.
+    - edges that appear twice are discarded. This allows representation of
+      polygons with holes in them.
+
```
```suggestion
    - Duplicate vertices are removed, as this can cause problems with
      triangulation algorithms
    - Duplicate edges are discarded. This allows representation of
      polygons with holes in them.

```

TimMonko commented on napari/layers/shapes/_accelerated_triangulate_dispatch.py:
```diff
@@ -256,6 +265,283 @@ def create_box_from_bounding_py(bounding_box: npt.NDArray) -> npt.NDArray:
+    CoordinateArray2D,
+    np.ndarray[tuple[int, Literal[2]], np.dtype[np.int64]],
+]:
+    """Get a list of edges that must be in triangulation for a path or polygon.
```
```suggestion
    """Normalize vertices and edges for a path or polygon.
```

TimMonko commented on napari/layers/shapes/_accelerated_triangulate_dispatch.py:
```diff
@@ -256,6 +265,283 @@ def create_box_from_bounding_py(bounding_box: npt.NDArray) -> npt.NDArray:
+        else:
+            edges.add(edge)
+
+    edges.remove((0, 0))
```
why does this need removed?

Czaki replied:
Because it is added in line 415 (`edge = (current_idx, prev_idx)` ) to keep simplicity of the code. Instead, the first turn of the loop needs to be implemented separately.

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
at this point, I'm definitely running out of steam in my review, so keep that in mind if this one is more sparse

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
 
 
-def triangulate_face(data: npt.NDArray) -> tuple[npt.NDArray, npt.NDArray]:
+def _cull_triangles_not_in_poly(vertices, triangles, poly):
```
the name and docstring of this function isn't helping me understand its purpose.

Is it to remove any triangles that would ... I'm not really sure I know the word, fill in the convex hull that should otherwise be convex shape edges?

jni replied:
Clarified in docstring

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
+    Parameters
+    ----------
+    vertices: np.ndarray[np.floating], shape (N, 2)
+        The vertices of the triangulation. Holes in the polygon are defined by
```
This repeats with `poly` description. I'm thinking the second sentence here is not supposed to be here?

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
+    Returns
+    -------
+    culled_triangles: np.ndarray[np.intp], shape (P, 3), P ≤ M
+        A subset of the input triangles.
```
```suggestion
        Array of triangles with centroids inside the polygon.
```

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
 
-def triangulate_face(data: npt.NDArray) -> tuple[npt.NDArray, npt.NDArray]:
+def _cull_triangles_not_in_poly(vertices, triangles, poly):
+    """Remove triangles that are not inside the polygon.
```
```suggestion
    """Remove triangles with centroids that are not inside the polygon.
```

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
+        The vertices of the triangulation. Holes in the polygon are defined by
+        an embedded polygon that starts from an arbitrary point in the
+        enclosing polygon and wind in the opposite direction.
+    axis: int
```
```suggestion
        The vertices of the triangulation.
    axis: int
```

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
+def _fix_vertices_if_needed(
+    vertices: CoordinateArray2D, axis: int | None, value: float | None
+) -> CoordinateArray:
+    """Fix the vertices if they are not planar along the given axis.
```
```suggestion
    """Ensure vertices are planar along a given axis.
    
    If an axis and value are provided, this function inserts the value as
    a new coordinate along the given axis. 
    Used to convert 2D vertices in 3D vertices along the plane.
```

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
+        Tuple of vertices, offsets, and triangles of the face.
+    edges : tuple[np.ndarray, np.ndarray, np.ndarray]
+        Tuple of vertices and triangles of the edges.
+    """
```
```suggestion
    face : tuple[np.ndarray, np.ndarray]
        Tuple of vertices and triangles of the edges.
    edges : tuple[np.ndarray, np.ndarray, np.ndarray]
        Tuple of vertices, offsets, and triangles of the face.
    """
```

TimMonko commented on napari/layers/shapes/_shapes_utils.py:
```diff
@@ -682,12 +590,186 @@ def triangulate_ellipse(
+    return new_vertices
+
+
+def triangulate_face_and_edges(
```
this function is used enough that I think the body would benefit from more comments. In particular its not clear to me `if _triangulate_dispatch.is_convex(data2d):` why its used. Is it because its just a faster implementation for simple polygons?

Czaki replied:
If the polygon is convex, then you could use the linear ($O(n)$) method of polygon triangulation instead of $O(n \log(n))$

TimMonko reviewed (commented):
Overall, I've done my best to review, mostly focusing on improving docstrings and comments. 

I ran completely out of steam on _shapes_utils which is unfortunate because its like 1/3rd of the PR. If there is something in there fore me to look at specifically I can. 

I'll try to circle back to the public functions once I have more time and energy.

I could approve this PR if needed, but I can't say I'm like an expert on it or anything, mostly I can just get the theme that we are recalculating triangulation in many ways to fix doubled things like edges/vertices, remove triangles with centroids outside polygons (and thus in holes), and improving calculations for 2D -> 3D shapes.

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -2,8 +2,29 @@
+
+
+class Orientation(Enum):
+    """Orientation of a triangle."""
```
```suggestion
    """Orientation of a triangle.
    
    The terms assume napari's preferred coordinate frame, in which
    the 0th axis, y, is pointing down, and the 1st axis, x, is pointing
    right. If one of the axes is flipped, the observed orientation
    would also flip.
    """
```

jni committed:
Improved description for convexity check

Co-authored-by: Tim Monko <timmonko@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni committed:
Clarify that orientation is defined in napari's default axes

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni committed:
Update return docstring for is_convex

Co-authored-by: Tim Monko <timmonko@gmail.com>

Czaki committed:
Apply suggestions from code review

Co-authored-by: Tim Monko <timmonko@gmail.com>

Czaki committed:
add comment with explanation

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -633,3 +775,182 @@ def create_box_from_bounding(bounding_box: np.ndarray) -> np.ndarray:
+@njit(cache=True)
+def reconstruct_polygons_from_edges(
+    vertices: CoordinateArray, edges: EdgeArray
+) -> list[CoordinateArray2D] | list[CoordinateArray3D]:
```
@Czaki ~~two~~ **edit** three 😅  things here:

1. (most important) Is this expected to include edges connecting inner and outer polygons? The docstring should explicitly specify whether "inter-polygon" edges are included and removed by the algorithm, or whether they should already be removed.
2. is this just connected components or is it something else? (also good to include in docstring)
3. (mostly just me trying to understand) I was trying to figure out what numba.typed.List was and found [this thread](https://numba.discourse.group/t/performance-of-typed-list-outside-of-jit-functions/2560/21) which shows that it's very slow outside of njit functions. So (a) do these lists every make it out of numba? (I will try to follow the code but wanted to write down my thoughts) (b) if we are creating the lists within numba and all the computation is within numba, is there a performance advantage? (ie would numba just promote a 'list' declaration to a typed list internally?)

Czaki replied:
1. improved docstring 
2. improved docstring
3. It looks like code compiled without `numba.typed.List`, so removed. However, this code is never triggered without numba.

jni commented on napari/layers/shapes/_accelerated_triangulate.py:
```diff
@@ -633,3 +775,182 @@ def create_box_from_bounding(bounding_box: np.ndarray) -> np.ndarray:
+                    # In case of an open chain, exit.
+                    break
+            polygons.append([vertices[x] for x in poly])
+    return polygons
```
I'm having a hard time following this algorithm. I'm wondering if we can avoid the breaks inside nested if statements, which make it difficult to understand to what level of nesting we are breaking out. One small suggestion is changing:

```python
for i in range(n_edges):
    if not used[i]:
        # huge loop body
```

to

```python
for i in range(n_edges):
    if used[i]:
        continue
    # huge loop body
```

Same for the incident edges loop.

I'm also wondering about "found" vs "closed". It seems to me that open chains are just ignored? In what situations can this algorithm encounter open chains?

jni replied:
Actually changing the nesting locally made it clear — open chains are just added anyway?

Again, maybe none of this would be surprising if I understood what this function is supposed to do... 😅

jni committed:
Reduce nesting by using continue

jni committed:
Rename used to visited (more common)

Czaki committed:
improve documentaion.

Czaki committed:
fix typing

jni committed:
Reorient SA polygon, reference SA in array name

Czaki committed:
Merge branch 'main' into polyhole

jni committed:
Add comments clarifying certain code paths

jni committed:
Clarify docstring of cull faces function

jni committed:
Improve docstring for fix vertices

Co-authored-by: Tim Monko <timmonko@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jni committed:
fix docstring of triangulate_face_and_edges

Co-authored-by: Tim Monko <timmonko@gmail.com>

jni committed:
Mild refactor of main triangulation function

jni committed:
Comment and docstring updates

jni committed:
Remove outdated unused code

jni committed:
Remove outdated unused code
